lib/rich-text/html.rb

```diff
@@ -1,66 +1,195 @@
+          end
+
+        # Object insert
+        elsif op.value.is_a?(Hash) && key = @inline_formats.keys.detect { |k| op.value.key?(k) }
+          format = @inline_formats[key.to_sym]
+          new_block_format = format[:block_format] if format.key?(:block_format)
+          els << apply_format(format, op.value, op)
+        end
+
+        # Flush the element flow when switching block formats
+        if new_block_format != current_block_format
+          render_block(ops.last, els.take(els.length - 1), current_block_format)
+          current_block_format = new_block_format
+          els.last(1)
         else
-          line = apply_tag(@default_block_tag, line, true)
+          els
         end
       end
-      line
+
+      # Render a block wrapper for the inlined elements
+      # places the resulting block into the root document
+      render_block(ops.last, elements, current_block_format)
     end
 
-    def apply_tags(tags, text, attributes)
-      attributes.inject(text) do |content, (key, value)|
-        apply_tag(tags[key], content, value)
+    # renders a block for a collection of elements based on a final operation
+    def render_block(op, elements, default_block_format=@default_block_format)
+      return unless elements.any?
+
+      # manually build block attributes to normalize malformed structures
+      block_attrs = op.attributes.slice(*@block_formats.keys)
+
+      # direct insertions (like "hr" tags) omit block format entirely
+      # install these elements directly into the document
+      if !default_block_format
+        # remove tag formats from these insertions so that they don't get wrapped
+        block_attrs = Hash[block_attrs.reject { |k, v| @block_formats[k.to_sym].key?(:tag) }]
+        return elements.each do |el|
+          el = apply_formats(@block_formats, el, op, attributes: block_attrs)
+          @root.add_child(el)
+        end
+      end
+
+      # reject any inline formats, and assure a block tag format is defined
+      unless block_attrs.detect { |k, v| @block_formats[k.to_sym].key?(:tag) }
+        unless @block_formats.key?(default_block_format.to_sym)
+          raise TypeError.new("block format #{default_block_format} is not defined")
+        end
+        block_attrs[default_block_format.to_sym] = true
       end
+
+      el = apply_formats(@block_formats, elements, op, attributes: block_attrs)
+      @root.add_child(el)
+    end
+
+    def apply_formats(formats, content, op, attributes:nil)
+      attributes ||= op.attributes
+
+      # order of operations for rendering formats
+      # tag formats are applied first (sorted by priority)
+      # other attribute formats follow (sorted by priority)
+      ordered_formats = attributes.keys.map {|k| formats[k.to_sym] }.compact.sort do |a, b|
+        a = [a.key?(:tag) ? 0 : 1, a[:priority] || Float::INFINITY]
+        b = [b.key?(:tag) ? 0 : 1, b[:priority] || Float::INFINITY]
+        a <=> b
+      end
```

Comment:
How did you settle on the default ordering rule for applying formats with tags first? Would things work out OK if we simply applied formats in alphabetical order by key name, instead of this priority system? Doesn't look like the default config uses any `:priority` values...

Reply:
Order is important when evaluating tag versus attribute formats (or – "other" formats that just apply an element modifier). Tag formats need to be rendered first so that there's something there for the attribute/other format to target. Priority is just an added courtesy that would allow tag order to be weighted if you cared about such things – I don't have a specific case where you would, but it seems plausible.